Use 70% less memory for TokenizedLines by using a different representation

nathansobo opened the pull request:
Previously, we naively stored every token on every line as its own object, and made things even worse by storing an array of scopes on every token. This caused `TokenizedLine`s to consume much more memory than is strictly necessary.

This PR switches `TokenizedLine` to a much more compact (but also more cryptic) representation. Instead of repeating the scopes on each token, we instead represent the entire tokenized line as a _nested word_, which intermingles start and end tags with text much like HTML.

Consider the following JavaScript fragment:

``` js
function() {}
```

Previously, we would store this as an array of tokens, each with an array of its containing scopes:

``` js
[
  {value: 'function', scopes: ['source.js', 'meta.function.js', 'storage.type.function.js']},
  {value: ' ', scopes: ['source.js', 'meta.function.js']},
  {value: '(', scopes: ['source.js', 'meta.function.js', 'punctuation.definition.parameters.begin.js']},
  {value: ')', scopes: ['source.js', 'meta.function.js', 'punctuation.definition.parameters.end.js']},
  {value: ' ', scopes: ['source.js', 'meta.function.js']},
  {value: '{', scopes: ['source.js', 'punctuation.section.scope.begin.js']},
  {value: '}', scopes: ['source.js', 'punctuation.section.scope.end.js']}
]
```

Now this is stored more like HTML, with strings intermingled with scope start / end tags.

``` js
// Note: This JSON is just for example purposes. Read below for the actual representation.

[
  {start: 'source.js'},
    {start: 'meta.function.js'},
      {start: 'storage.type.function.js'},
        "function",
      {end: 'storage.type.function.js'},
      " ",
      {start: 'punctuation.definition.parameters.begin.js'},
        "(",
      {end: 'punctuation.definition.parameters.begin.js'},
      {start: 'punctuation.definition.parameters.end.js'},
        ")",
      {end: 'punctuation.definition.parameters.end.js'},
    {end: 'meta.function.js'},
    " ",
    {start: 'punctuation.section.scope.begin.js'},
      "{",
    {end: 'punctuation.section.scope.begin.js'},
    {start: 'punctuation.section.scope.end.js'},
      "}",
    {end: 'punctuation.section.scope.end.js'},
  // note 'source.js' isn't closed yet
]
```

This may look more verbose, but it eliminates the need to repeat the same scopes on successive tokens in the same branch of the syntax tree. In actuality, the scopes are encoded as simple arrays of integers. Negative integers indicate start and end tags, with odd integers being start tags and even integers being end tags for the odd integer preceding them. Positive integers stand in for strings. The integers can be translated back to scope strings by consulting a lookup table on the grammar registry. We retain the original string and combine the two pieces of data as needed.

So in reality, the stored representation looks like this.

``` js
{
  text: "function() {}",
  tags: [-1, -3, -5, 8, -6, 1, -7, 1, -8, -9, 1, -10, -4, 1, -11, 1, -12, -13, 1, -14]
}
```
### Measurements

Here are some heap snapshots taken with `benchmark/fixtures/huge.js` in the Atom repository. This file is just a copy of jQuery 1.7.1, but has some good examples of lines with extremely complex scope structures.

First, on master:

![screenshot 2015-05-14 19 07 26](https://cloud.githubusercontent.com/assets/1789/7637260/e40779b6-fa6c-11e4-8441-ef42af5e8f9f.png)

`TokenizedLine` instances consume over 27MB. The largest instance is expanded in the screenshot, where you can see the vast majority of the memory for that line being consumed by the `tokens` array (~25KB). Repeat that for thousands of lines and you can see the problem.

Now, the same file on this branch:

![screenshot 2015-05-14 19 09 57](https://cloud.githubusercontent.com/assets/1789/7637303/40a52bdc-fa6d-11e4-88b3-e64479cccda3.png)

Still more memory than I'd like to consume, but much less, around 8.4MB. You can see that the `tags` array takes up much less space than the `tokens` array, but still more than it seems like it should. I can save more with typed arrays, but only about 500KB in my experiment.

JS objects still seem to impose a frustrating amount of overhead. Starting to wonder if we want to drop some of our low-level storage structures to C++, but I'd rather improve the structure of this layer before contemplating that. The `TokenizedLine` instances are also now much simpler objects, meaning they terminate in primitives much more quickly. I wonder if this helps the GC in that the heap is simpler and there are fewer objects to trace. This is just speculation however.
### Steps
- [x] Get core specs passing with new representation
- [x] Eliminate usages of the `tokens` array shim in core, which creates a lot of garbage
- [x] Post some heap snapshot comparisons
- [x] Publish a new major revision of first-mate and link it in package.json
- [x] Ensure all package specs pass
- [x] Switch hot package paths off the tokens shim

Commit:
WIP: Start integrating with nested words version of first-mate

Commit:
Adapt to first-mate returning purely numeric tag arrays

Commit:
Break out leading soft tabs after switching to numeric tag arrays

Commit:
Break out hard tabs and paired characters from numeric tag arrays

Commit:
Mark leading/trailing whitespace in new TokenizedLine representation

Commit:
Make TokenizedLine::softWrapAt work with new token representation

Commit:
Fix TokenizedLine::copy for new representation

Commit:
Delete spec that’s no longer relevant

Commit:
Substitute invisibles in initial scan

Commit:
Fix bufferRangeForScopeAtPosition with new tags array scheme

Commit:
Drop unused methods

Commit:
Update spec based on new interface for Grammar::tokenizeLine

Commit:
Fix soft wrapping with tags array, including soft-wrap indent

Commit:
Assign ::hasInvisibleCharacters on Token shims

Commit:
Fix assignment of last non-whitespace column for paired characters

Commit:
Assign ::isHardTab to shim tokens

Commit:
Don’t build ::specialTokens unless properties are provided

Commit:
Account for softWrapIndent when assigning indices on soft wrap line

Commit:
Don’t assume same token instances in spec now that we use shims

Commit:
decodeContent -> decodeTokens

benogle commented:
Are the scope name strings in a symbol table or something, or are the strings repeated?

nathansobo commented:
@benogle They're stored in a symbol table on the `GrammarRegistry`. In the `tags` arrays they are negative integers. Odd for start scopes, even for end scopes. Could have used one universal symbol for the end scope but I wanted to preserve a little information for debugability purposes. We're not going to create so many scopes that using two ids for each scope is a problem. Only the odd ids are actually stored in the symbol table.

Commit:
Rename parentScopes to openScopes

Commit:
Generate line HTML based on tags instead of tokens

This avoids creating a bunch of tokens as temporary objects since they
are no longer stored.

Commit:
Remove some dead code

Commit:
:art:

Commit:
Remove dead code

Commit:
Add a TokenIterator and use it for tokens shim

Commit:
Use TokenIterator for position translation

Commit:
Use TokenIterator to build line HTML

Commit:
Use TokenIterator to compute horizontal pixel positions

Instead of the TokenizedLine::tokens shim

Commit:
Switch character measurement to TokenIterator

Instead of using TokenizedLine::tokens shim

Commit:
Use TokenIterator in DisplayBuffer instead of tokens shim

Commit:
Don’t include tokens in presenter state for lines

Commit:
Remove more usages of tokens shim

Commit:
Avoid tokens shim in suggestedIndent code

Commit:
Avoid tokens shim in isBufferRowCommented

Commit:
Remove unused TokenizedBuffer methods that relied on tokens shim

Commit:
Merge branch 'master' into ns-less-memory-for-tokens

Conflicts:
	spec/tokenized-buffer-spec.coffee
	src/tokenized-buffer.coffee

nathansobo commented:
Published some heap screenshots and analysis.

Commit:
:arrow_up: first-mate to 4.0.0 for tagged token representation

Commit:
:fire: debugger

Commit:
Merge branch 'master' into ns-less-memory-for-tokens

Commit:
:arrow_up: first-mate

Commit:
:arrow_up: first-mate

Commit:
Return scopes at very end of token for the last token on a line

Commit:
Fix spurious lint error by rearranging syntax

Commit:
Use scope of last token if we iterate off the end

Commit:
:arrow_up: first-mate

nathansobo commented:
A couple failures left. One appears to be an actual bug in the previous version of the grammar where scopes weren't being properly propagated to nested capture rules. It's manifested in the language-shellscript tests. May need to find a workaround so the tests pass against both versions.

The other failure I can't yet reproduce locally and will need to investigate further.

Commit:
:arrow_up: language-shellscript for spec fixes against first-mate 4.x

Commit:
Merge remote-tracking branch 'origin/master' into ns-less-memory-for-tokens

grizzilus commented:
Not sure if this branch is ready for testing, but I just built it and have some with the `language-latex` package (which isn't a core package but still a pretty commonly used one).

If I open or create a file with the `LaTeX` grammar and start typing `\chapter*{test}` at the beginning of a line, when I reach the left brace I get the error `Uncaught Error: Encountered an invalid scope end id. Popped -53, expected to pop -59.`

Again, not sure if that is helpful, but I thought I'd mention it just to be safe.

nathansobo commented:
Thanks a lot @grizzilus. I'll definitely take a look at that. Did you build against a version following  66d9b30?

grizzilus commented:
@nathansobo This was against 3dcf9a4

Commit:
:arrow_up: first-mate

Commit:
Merge remote-tracking branch 'origin/master' into ns-less-memory-for-tokens

nathansobo commented:
@grizzilus Think today's changes to `first-mate` should resolve your issue.

nathansobo commented:
@atom/feedback It would be great if folks could build this branch and try this out, especially with a range of different language bundles and content. I'm worried about regressions due to my changes to the text-mate parser, though we do have quite a bit of test coverage.

grizzilus commented:
@nathansobo Just built adb4988 and it looks good, you definitely took care of that issue

lee-dohm commented:
I'll work with it tonight and let you know what I find.

lee-dohm commented:
I ran into this when using toggle-quotes on the CoffeeScript string: `"adding the GitHub events hook to #{user}/#{repo}"` When rebuilding using `master`, it did not reproduce.

**Atom Version**: 0.200.0-adb4988
**System**: Mac OS X 10.10.3
**Thrown From**: [toggle-quotes](https://github.com/atom/toggle-quotes) package, v0.11.0

### Stack Trace

Uncaught TypeError: Cannot read property 'split' of undefined

```
At /Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:683

TypeError: Cannot read property 'split' of undefined
  at ScopeMatcher.matches (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/first-mate/lib/scope-selector-matchers.js:50:28)
  at PathMatcher.matches (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/first-mate/lib/scope-selector-matchers.js:90:21)
  at ScopeSelector.module.exports.ScopeSelector.matches (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/first-mate/lib/scope-selector.js:15:27)
  at TokenizedBuffer.module.exports.TokenizedBuffer.bufferRangeForScopeAtPosition (/Applications/Atom.app/Contents/Resources/app.asar/src/tokenized-buffer.js:691:25)
  at DisplayBuffer.module.exports.DisplayBuffer.bufferRangeForScopeAtPosition (/Applications/Atom.app/Contents/Resources/app.asar/src/display-buffer.js:1066:35)
  at toggleQuoteAtPosition (/Users/Lee/.atom/packages/toggle-quotes/lib/toggle-quotes.coffee:10:32)
  at /Users/Lee/.atom/packages/toggle-quotes/lib/toggle-quotes.coffee:5:7
  at TextBuffer.module.exports.TextBuffer.transact (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:678:18)
  at TextEditor.module.exports.TextEditor.transact (/Applications/Atom.app/Contents/Resources/app.asar/src/text-editor.js:1132:26)
  at toggleQuotes (/Users/Lee/.atom/packages/toggle-quotes/lib/toggle-quotes.coffee:2:10)
  at atom-text-editor.<anonymous> (/Users/Lee/.atom/packages/toggle-quotes/lib/toggle-quotes.coffee:60:9)
  at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:238:29)
  at /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:3:61
  at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:519:16)
  at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:354:22)
  at HTMLDocument.module.exports.WindowEventHandler.onKeydown (/Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:178:20)

```

### Commands

```
  2x -3:34.8.0 editor:move-to-first-character-of-line (atom-text-editor.editor.is-focused)
     -3:33.6.0 core:move-down (atom-text-editor.editor.is-focused)
  7x -3:33.1.0 core:select-down (atom-text-editor.editor.is-focused)
     -3:31 core:backspace (atom-text-editor.editor.is-focused)
     -3:30.4.0 core:save (atom-text-editor.editor.is-focused)
     -1:08.9.0 editor:newline (atom-text-editor.editor.is-focused)
  2x -1:07.5.0 core:backspace (atom-text-editor.editor.is-focused.autocomplete-active)
  2x -0:51.1.0 editor:move-to-first-character-of-line (atom-text-editor.editor.is-focused.autocomplete-active)
     -0:50.3.0 editor:move-to-end-of-screen-line (atom-text-editor.editor.is-focused)
     -0:50 core:move-left (atom-text-editor.editor.is-focused)
  2x -0:47.2.0 editor:move-to-first-character-of-line (atom-text-editor.editor.is-focused)
     -0:46.6.0 core:move-down (atom-text-editor.editor.is-focused)
  6x -0:46.3.0 core:select-down (atom-text-editor.editor.is-focused)
     -0:45 core:backspace (atom-text-editor.editor.is-focused)
     -0:44.3.0 core:save (atom-text-editor.editor.is-focused)
  3x -0:04.8.0 toggle-quotes:toggle (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "followSymlinks": true,
    "disabledPackages": [
      "autocomplete-paths",
      "space-underscore"
    ],
    "projectHome": "/Users/Lee/Source",
    "themes": [
      "atom-dark-ui",
      "advanced-railscasts-syntax"
    ],
    "ignoredNames": [
      ".git",
      ".hg",
      ".svn",
      ".DS_Store",
      ".keep",
      "._*",
      "*.pyc"
    ]
  }
}
```

### Installed Packages

```coffee
# User
advanced-railscasts-syntax, v0.1.10
asciidoc-preview, v0.5.0
autocomplete-emojis, v2.2.2
block-travel, v1.0.2
bug-report, v0.7.0
file-type-icons, v0.6.0
indentation-indicator, v0.5.1
language-generic-config, v0.2.0
language-haml, v0.20.0
language-r, v0.2.0
language-swift, v0.4.0
language-todotxt, v0.6.1
language-treetop, v0.1.1
linter, v0.12.3
linter-coffeelint, v0.2.3
linter-csslint, v0.0.12
linter-less, v0.4.0
linter-rubocop, v0.2.6
linter-scss-lint, v0.0.14
linter-shellcheck, v0.0.7
minimap, v4.9.0
package-sync, v0.2.0
pigments, v0.3.0
red-wavy-underline, v0.3.0
set-syntax, v0.3.0
soft-wrap-indicator, v0.7.0
tabs-to-spaces, v0.10.0
toggle-quotes, v0.11.0
white-cursor, v0.6.0
word-jumper, v0.2.0

# Dev
tabs-to-spaces, v0.10.0
```

nathansobo commented:
@lee-dohm Can you reproduce this reliably? I'm having trouble.

lee-dohm commented:
@nathansobo I could until I built `master` and then went back to your branch. Seeing what I can do to reproduce it better.

nathansobo commented:
I'll try to reason through it from the stack trace in the meantime.

nathansobo commented:
Think I *may* have found the issue.

lee-dohm commented:
I built clean and can't reproduce the issue again :disappointed:

Commit:
:arrow_up: first-mate

nathansobo commented:
@lee-dohm I figured it out. Previously you could only convert tag integers to scope names with the start tag. Now I've made it work with both the start tag (an odd negative integer) and the end tag (an even negative integer). Maybe you could build again on latest and hunt for any other problems? Thanks!!!

lee-dohm commented:
Building latest and I'll keep poking around.

Commit:
Avoid error in donna npm while generating documentation

Commit:
Merge remote-tracking branch 'origin/master' into ns-less-memory-for-tokens

Commit:
Add TokenizedLine::getTokenIterator

This will not be part of the public API but will replace another
non-public API usage in autocomplete-plus.

maxbrunsfeld commented on src/token-iterator.coffee:
```diff
@@ -0,0 +1,85 @@
+  isAtomic: ->
+    @isSoftTab() or @isHardTab() or @isSoftWrapIndentation() or @isPairedCharacter()
+
+TokenIterator.instance = new TokenIterator
```
It looks like this is here to save on object allocations. What do you think about caching a `TokenIterator` on `TokenizedBuffer` or `DisplayBuffer` instead of as a global?

nathansobo replied:
Yeah that could work. I was trying to avoid adding yet another reference that had to be stored on `TokenizedLine`. But it's probably not that big of a deal.

maxbrunsfeld replied:
> I was trying to avoid adding yet another reference that had to be stored on TokenizedLine.

Yeah, I agree that that's very worthwhile. I was just wondering if it could be cached at a higher level. `TokenizedLine` only uses the `TokenIterator` if a package calls `::getTokenIterator`, right? In that case, could we just instantiate a fresh one?

nathansobo replied:
I went ahead and added the reference. We need the iterator within the `TokenizedLine` line as well  to support different use cases, but one reference per line isn't a huge deal considering the other stuff we store.

Commit:
Pass TokenIterator reference instead of using global singleton

Adds an extra reference to each tokenized line but is also more sane.

/cc @maxbrunsfeld

maxbrunsfeld commented:
:fireworks:

winnieXY commented:
@nathansobo : I'm sadly still able to reproduce the error with latex files. I'm getting a 
Uncaught Error: Encountered an invalid scope end id. Popped -13, expected to pop -71 Every time I open a .tex file.

I'm currently running a build from the git. (commit: 75d02cdbc92b4ed5ef0139a4abe24fddd8d0e397). Any ideas what is going wrong here?

as-cii commented:
> I'm sadly still able to reproduce the error with latex files. I'm getting a 
> Uncaught Error: Encountered an invalid scope end id. Popped -13, expected to pop -71 Every time I open a .tex file.

Thanks for reporting this issue, @winnieXY: your help is much appreciated! :bow:

Are you able to reproduce this consistently? If so, could you attach the steps here, please? Any clue to make this behavior repeatable would be really, really helpful!

Thanks!

winnieXY commented:
Yes, I'm able to reproduce this consistently in the tex files of my PhD thesis ;(
I'm failing currently to write a minimal working example.

winnieXY commented:
/usr/share/atom/resources/app.asar/src/tokenized-buffer.js:573

Error: Encountered an invalid scope end id. Popped -13, expected to pop -65.
  at TokenizedBuffer.module.exports.TokenizedBuffer.scopesFromTags (/usr/share/atom/resources/app.asar/src/tokenized-buffer.js:573:21)
  at TokenizedBuffer.module.exports.TokenizedBuffer.openScopesForRow (/usr/share/atom/resources/app.asar/src/tokenized-buffer.js:555:21)
  at TokenizedBuffer.module.exports.TokenizedBuffer.tokenizeNextChunk (/usr/share/atom/resources/app.asar/src/tokenized-buffer.js:286:105)
  at /usr/share/atom/resources/app.asar/src/tokenized-buffer.js:263:26
  at /usr/share/atom/resources/app.asar/node_modules/underscore-plus/node_modules/underscore/underscore.js:666:47

winnieXY commented:
I found another very annoying bug. Sometimes the cursor goes mad and I'm typing somewhere else than the cursor is. this is easy reproduceable with soft wrap lines and wrap lines at 80 chars. If I type a sentence and add at the end a \textsc{ which is then automatically wrapped into the next line, the cursor is not in the position it should be.

winnieXY commented:
This is easy reproduceable with a very small .tex file. Do you have an emailadress than I can send you the minimal working example.

mnquintana commented:
@winnieXY You can always paste the example into a Gist

nathansobo commented:
There's something I missed in the changes to `first-mate`. Please do post an example or point at an example repo. @atom/feedback if anyone can mess around with LaTeX files and get a reproduction case, that would be great. @kevinsawicki release is going to be blocked until we can iron this out.

mnquintana commented:
@winnieXY What LaTeX package(s) do you have installed?

winnieXY commented:
autocomplete-bibtex, latex and language-latex - nothing else.There are two errors. First the strange behaviour of the cursor, second the error message popping up. The second one is not reproduceable with a minimal working example, thus I thought about sending one chapter of my PhD thesis to the person working on the fix - however I do not want to make this public (as this is still very alpha and contains results which are not yet published).

winnieXY commented:
The first error is already reported here: #6885. Maybe it is not related to this change, I don't know

nathansobo commented:
Okay, I've reverted the merge. Seems like there were at least a couple of issues I didn't detect. Thanks for the rapid feedback on these. @winnieXY can you email me the example that reproduces the issue at `nathan@github.com`. Thanks again guys. Sorry for the history pollution but this change was a tough one to predict every corner case.

ypresto commented on spec/tokenized-buffer-spec.coffee:
```diff
@@ -580,7 +572,7 @@ describe "TokenizedBuffer", ->
     describe "when the selector matches a run of multiple tokens at the position", ->
       it "returns the range covered by all contigous tokens (within a single line)", ->
-        expect(tokenizedBuffer.bufferRangeForScopeAtPosition('.function', [1, 18])).toEqual [[1, 6], [1, 28]]
+        expect(tokenizedBuffer.bufferRangeForScopeAtPosition('.meta.function', [1, 18])).toEqual [[1, 6], [1, 28]]
```
I think this is destructive change on selector if this doesn't accept `.function` selector anymore and we should use `.meta.function` instead.

This caused bug on the plugin because it uses `.regexp` instead of `.string.regexp`.
https://github.com/klorenz/atom-regex-railroad-diagrams/issues/42

ypresto replied:
Merge commit 0cd1f110b581e4c29fa4cec432cdbb93ecba2409 reproduces bug, and it's parent d75d202d337142e206e500c10fed34151b4cc750 does not reproduce.

nathansobo replied:
@ypresto You're right. Sorry about that. I switched to using `ScopeSelector` from `first-mate` without recalling that their matching criteria is stricter. Can you switch to `.string.regexp` until I can get a fix in today?

ypresto replied:
OK I already PRed that repo ;).

I think it'd better to provide backward compatibility (with deprecation warning) only if dot prefix is used.
